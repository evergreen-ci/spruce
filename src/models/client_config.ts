import { plainToClass } from "class-transformer";

export function ConvertToClientConfig(raw: string): ClientConfig {
  return plainToClass(ClientConfig, JSON.parse(raw) as object);
}

/* tslint:disable:variable-name */
export class ClientConfig {
  public api_url: string;
  public ui_url: string;

  public user?: string;
  public api_key?: string;
}

export function IsValidConfig(toCheck: object): boolean {
  const config = toCheck as ClientConfig;
  return config.api_url !== undefined && config.ui_url !== undefined;
}