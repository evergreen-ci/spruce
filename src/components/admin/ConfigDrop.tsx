import { DropzoneArea } from 'material-ui-dropzone'
import * as React from 'react';
import {ClientConfig, ConvertToClientConfig} from '../../models/client_config';
import '../../styles.css';

class Props {
  public updateClientConfig: (configObj: ClientConfig) => void;
}

interface State {
  processing: boolean,
  error: string
}

export class ConfigDrop extends React.Component<Props, State> {

  constructor(props: Props) {
    super(props);
    this.state = {
      processing: false,
      error: null
    };
  }

  public render() {
    return (
      <DropzoneArea onChange={this.handleDropAreaChange} 
        dropzoneText={"Drop your config file here"} 
        showAlerts={false} 
        filesLimit={1}
        acceptedFiles={["application/json"]}/>
    )
  }

  private handleDropAreaChange = (fileArray: File[]) => {
    if(fileArray.length !== 0) {
      this.setState({ 
        processing: true,
      });
      this.upload(fileArray[0]);
    }
  }


  private upload = (file: File) => {
    const reader = new FileReader();
    reader.onload = () => {
      const raw = reader.result.toString();
      const configObj = ConvertToClientConfig(raw);
      this.props.updateClientConfig(configObj);
    };
    reader.readAsText(file);
    
  }
}

export default ConfigDrop;