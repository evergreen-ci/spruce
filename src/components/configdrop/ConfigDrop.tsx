import { Button, Card, CardActions, CardContent, IconButton, Snackbar } from '@material-ui/core';
import CloseIcon from '@material-ui/icons/Close';
import { DropzoneArea } from 'material-ui-dropzone'
import * as React from 'react';
import { ClientConfig, ConvertToClientConfig } from '../../models/client_config';
import '../../styles.css';

class Props {
  public updateClientConfig: (configObj: ClientConfig) => void;
}

interface State {
  snackbarOpen: boolean,
  snackbarMessage: string,
  error: string
  newConfig: ClientConfig,
}

export class ConfigDrop extends React.Component<Props, State> {

  constructor(props: Props) {
    super(props);
    this.state = {
      snackbarOpen: false,
      snackbarMessage: "",
      error: null,
      newConfig: null
    };
  }

  public render() {
    return (
      <div>
        <Card>
          <CardContent>
            <DropzoneArea onChange={this.handleDropAreaChange}
              dropzoneText={"Drop your config file here"}
              filesLimit={1}
              acceptedFiles={["application/json"]} />
          </CardContent>
          <CardActions>
            <Button variant="outlined" className="save" onClick={this.save()}>Save</Button>
          </CardActions>
        </Card>
        <Snackbar open={this.state.snackbarOpen} message={this.state.snackbarMessage}
          anchorOrigin={{ vertical: "bottom", horizontal: "left", }} onClose={this.onCloseSnackbar}
          action={[<IconButton key="close" color="inherit" onClick={this.onCloseSnackbar}><CloseIcon /></IconButton>,]} />
      </div>
    )
  }

  private handleDropAreaChange = (fileArray: File[]) => {
    if (fileArray.length !== 0) {
      this.readConfig(fileArray[0]);
    } else {
      this.setState({
        newConfig: null
      });
    }
  }

  private readConfig = (file: File) => {
    const reader = new FileReader();
    reader.onerror = () => {
      this.setState({
        snackbarMessage: "Failed to read config file with error " + reader.error,
        snackbarOpen: true
      });
    };
    reader.onload = () => {
      const raw = reader.result.toString();
      const configObj = ConvertToClientConfig(raw);
      if (configObj.hasOwnProperty("user") && configObj.hasOwnProperty("api_key") &&
        configObj.hasOwnProperty("ui_url") && configObj.hasOwnProperty("api_url")) {
        this.setState({
          newConfig: configObj,
        });
      } else {
        this.setState({
          snackbarMessage: "Config file does not contain all required properties.",
          snackbarOpen: true
        });
      }
    };
    reader.readAsText(file);
  }

  private save = () => () => {
    if (this.state.newConfig !== null) {
      this.props.updateClientConfig(this.state.newConfig);
      this.setState({
        snackbarMessage: "Config file saved.",
        snackbarOpen: true
      });
    } else {
      this.setState({
        snackbarMessage: "Please upload a valid config file before saving.",
        snackbarOpen: true
      });
    }
  }

  private onCloseSnackbar = (event: React.SyntheticEvent | React.MouseEvent, reason?: string) => {
    this.setState({
      snackbarOpen: false
    });
  }
}

export default ConfigDrop;