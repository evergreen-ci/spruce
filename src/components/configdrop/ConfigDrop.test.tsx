import { Button } from '@material-ui/core';
import * as enzyme from "enzyme";
import { DropzoneArea } from 'material-ui-dropzone';
import * as React from "react";
import { ClientConfig } from '../../models/client_config';
import { ConfigDrop } from "./ConfigDrop";

describe("ConfigDrop", () => {

  const verifyValidConfig = jest.fn((configObj: ClientConfig) => {
    if (configObj !== undefined) {
      expect(configObj.user).toBe("some.user");
      expect(configObj.api_key).toBe("aczq3eG0gAgEV9uGZN5gCos1utnZPXIg");
      expect(configObj.api_url).toBe("https://evergreen.mongodb.com/api");
      expect(configObj.ui_url).toBe("https://evergreen.mongodb.com");
    }
  });

  const validWrapper = enzyme.mount(<ConfigDrop updateClientConfig={verifyValidConfig} />);
  const invalidWrapper = enzyme.mount(<ConfigDrop updateClientConfig={jest.fn()} />);

  it("matches snapshot", () => {
    // note: using enzyme.mount (instead of enzyme.shallow) causes the test suite to hang, so this test needs its own shallow wrapper
    const shallowWrapper = enzyme.shallow(<ConfigDrop updateClientConfig={jest.fn()} />);
    expect(shallowWrapper).toMatchSnapshot();
  })

  it("check config state object is not updated with an empty file", async () => {
    expect(invalidWrapper.state("newConfig")).toBeNull();
    const emptyFile = new File(["{}"], "config.json");
    const drop = invalidWrapper.find(DropzoneArea);
    // const save = invalidWrapper.find(Button);
    drop.simulate("change", { target: { files: [emptyFile] } });
    // save.simulate("click");
    expect(invalidWrapper.state("newConfig")).toBeNull();
    expect(invalidWrapper.state("snackbarOpen")).toBe(true);
    expect(invalidWrapper.state("snackbarMessage")).toBe("Config file does not contain all required properties.");
  })

  it("check config state object is not updated with an invalid file", () => {
    expect(invalidWrapper.state("newConfig")).toBeNull();
    const invalidJSON = {
      "user": "some.user",
      "api_url": "https://evergreen.mongodb.com/api",
    };
    const invalidFile = new File([JSON.stringify(invalidJSON)], "config.json");
    const drop = invalidWrapper.find(DropzoneArea);
    //const save = invalidWrapper.find(Button);
    drop.simulate("change", { target: { files: [invalidFile] } });
    //save.simulate("click");
    expect(invalidWrapper.state("newConfig")).toBeNull();
    expect(invalidWrapper.state("snackbarOpen")).toBe(true);
    expect(invalidWrapper.state("snackbarMessage")).toBe("Config file does not contain all required properties.");
  })

  it("check config state object is updated with a valid file", () => {
    expect(validWrapper.state("newConfig")).toBeNull();
    const validJSON = {
      "user": "some.user",
      "api_key": "aczq3eG0gAgEV9uGZN5gCos1utnZPXIg",
      "api_url": "https://evergreen.mongodb.com/api",
      "ui_url": "https://evergreen.mongodb.com"
    };
    const validFile = new File([JSON.stringify(validJSON)], "config.json");
    const drop = validWrapper.find(DropzoneArea);
    const save = validWrapper.find(Button);
    drop.simulate("change", { target: { files: [validFile] } });
    save.simulate("click");
    expect(verifyValidConfig).toHaveBeenCalled();
    expect(validWrapper.state("snackbarOpen")).toBe(true);
    expect(validWrapper.state("snackbarMessage")).toBe("Config file saved.");
  })
})