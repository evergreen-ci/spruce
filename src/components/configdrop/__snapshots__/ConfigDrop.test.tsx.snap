// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfigDrop matches snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ConfigDrop
    updateClientConfig={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <WithStyles(Card)>
          <WithStyles(CardContent)>
            <WithStyles(DropzoneArea)
              acceptedFiles={
                Array [
                  "application/json",
                ]
              }
              dropzoneText="Drop your config file here"
              filesLimit={1}
              onChange={[Function]}
            />
          </WithStyles(CardContent)>
          <WithStyles(CardActions)>
            <WithStyles(Button)
              className="save"
              onClick={[Function]}
              variant="outlined"
            >
              Save
            </WithStyles(Button)>
          </WithStyles(CardActions)>
        </WithStyles(Card)>,
        <WithStyles(Snackbar)
          action={
            Array [
              <WithStyles(IconButton)
                color="inherit"
                onClick={[Function]}
              >
                <pure(CloseIcon) />
              </WithStyles(IconButton)>,
            ]
          }
          anchorOrigin={
            Object {
              "horizontal": "left",
              "vertical": "bottom",
            }
          }
          message=""
          onClose={[Function]}
          open={false}
        />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <WithStyles(CardContent)>
              <WithStyles(DropzoneArea)
                acceptedFiles={
                  Array [
                    "application/json",
                  ]
                }
                dropzoneText="Drop your config file here"
                filesLimit={1}
                onChange={[Function]}
              />
            </WithStyles(CardContent)>,
            <WithStyles(CardActions)>
              <WithStyles(Button)
                className="save"
                onClick={[Function]}
                variant="outlined"
              >
                Save
              </WithStyles(Button)>
            </WithStyles(CardActions)>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": <WithStyles(DropzoneArea)
                acceptedFiles={
                  Array [
                    "application/json",
                  ]
                }
                dropzoneText="Drop your config file here"
                filesLimit={1}
                onChange={[Function]}
              />,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "acceptedFiles": Array [
                  "application/json",
                ],
                "dropzoneText": "Drop your config file here",
                "filesLimit": 1,
                "onChange": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": <WithStyles(Button)
                className="save"
                onClick={[Function]}
                variant="outlined"
              >
                Save
              </WithStyles(Button)>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": "Save",
                "className": "save",
                "onClick": [Function],
                "variant": "outlined",
              },
              "ref": null,
              "rendered": "Save",
              "type": [Function],
            },
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "action": Array [
            <WithStyles(IconButton)
              color="inherit"
              onClick={[Function]}
            >
              <pure(CloseIcon) />
            </WithStyles(IconButton)>,
          ],
          "anchorOrigin": Object {
            "horizontal": "left",
            "vertical": "bottom",
          },
          "message": "",
          "onClose": [Function],
          "open": false,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <WithStyles(Card)>
            <WithStyles(CardContent)>
              <WithStyles(DropzoneArea)
                acceptedFiles={
                  Array [
                    "application/json",
                  ]
                }
                dropzoneText="Drop your config file here"
                filesLimit={1}
                onChange={[Function]}
              />
            </WithStyles(CardContent)>
            <WithStyles(CardActions)>
              <WithStyles(Button)
                className="save"
                onClick={[Function]}
                variant="outlined"
              >
                Save
              </WithStyles(Button)>
            </WithStyles(CardActions)>
          </WithStyles(Card)>,
          <WithStyles(Snackbar)
            action={
              Array [
                <WithStyles(IconButton)
                  color="inherit"
                  onClick={[Function]}
                >
                  <pure(CloseIcon) />
                </WithStyles(IconButton)>,
              ]
            }
            anchorOrigin={
              Object {
                "horizontal": "left",
                "vertical": "bottom",
              }
            }
            message=""
            onClose={[Function]}
            open={false}
          />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <WithStyles(CardContent)>
                <WithStyles(DropzoneArea)
                  acceptedFiles={
                    Array [
                      "application/json",
                    ]
                  }
                  dropzoneText="Drop your config file here"
                  filesLimit={1}
                  onChange={[Function]}
                />
              </WithStyles(CardContent)>,
              <WithStyles(CardActions)>
                <WithStyles(Button)
                  className="save"
                  onClick={[Function]}
                  variant="outlined"
                >
                  Save
                </WithStyles(Button)>
              </WithStyles(CardActions)>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": <WithStyles(DropzoneArea)
                  acceptedFiles={
                    Array [
                      "application/json",
                    ]
                  }
                  dropzoneText="Drop your config file here"
                  filesLimit={1}
                  onChange={[Function]}
                />,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "acceptedFiles": Array [
                    "application/json",
                  ],
                  "dropzoneText": "Drop your config file here",
                  "filesLimit": 1,
                  "onChange": [Function],
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": <WithStyles(Button)
                  className="save"
                  onClick={[Function]}
                  variant="outlined"
                >
                  Save
                </WithStyles(Button)>,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "children": "Save",
                  "className": "save",
                  "onClick": [Function],
                  "variant": "outlined",
                },
                "ref": null,
                "rendered": "Save",
                "type": [Function],
              },
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "action": Array [
              <WithStyles(IconButton)
                color="inherit"
                onClick={[Function]}
              >
                <pure(CloseIcon) />
              </WithStyles(IconButton)>,
            ],
            "anchorOrigin": Object {
              "horizontal": "left",
              "vertical": "bottom",
            },
            "message": "",
            "onClose": [Function],
            "open": false,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;
